File: python/src/mpeg_o/access_policy.py
"""``AccessPolicy`` — who-may-decrypt-what policy dictionary."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True, slots=True)
class AccessPolicy:
    """Access policy describing who may decrypt which streams in an
    ``.mpgo`` file.

    Stored as a JSON string under ``/protection/access_policies`` on
    disk, so the policy is human-inspectable and recoverable
    independently of any key-management system.

    The policy is intentionally schema-free at this layer: the dict
    holds arbitrary key/value pairs that the application is free to
    interpret (typical fields: ``subjects``, ``streams``, ``expiry``,
    ``key_id``, ``audit_contact``).

    Parameters
    ----------
    policy : dict[str, Any], default {}
        Arbitrary key/value policy payload.

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: ``MPGOAccessPolicy`` · Java:
    ``com.dtwthalion.mpgo.protection.AccessPolicy``.
    """

    policy: dict[str, Any] = field(default_factory=dict)

File: java/src/main/java/com/dtwthalion/ttio/protection/AccessPolicy.java
/* TTI-O Java Implementation / Copyright (C) 2026 DTW-Thalion / SPDX-License-Identifier: LGPL-3.0-or-later */
package com.dtwthalion.ttio.protection;

import java.util.Map;

/**
 * Access policy describing who may decrypt which streams in an
 * {@code .tio} file.
 *
 * <p>Stored as a JSON string under
 * {@code /protection/access_policies} on disk, so the policy is
 * human-inspectable and recoverable independently of any
 * key-management system.</p>
 *
 * <p>Schema-free at this layer: the map holds arbitrary key/value
 * pairs the application interprets (typical fields: {@code subjects},
 * {@code streams}, {@code expiry}, {@code key_id},
 * {@code audit_contact}).</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOAccessPolicy}, Python
 * {@code ttio.access_policy.AccessPolicy}.</p>
 *
 * @param policy Arbitrary key/value policy payload.
 * @since 0.6
 */
public record AccessPolicy(Map<String, Object> policy) {
    /**
     * Defensive copy; substitutes empty map for null.
     *
     * @param policy Arbitrary key/value policy payload (null treated as empty).
     */
    public AccessPolicy {
        policy = policy != null ? Map.copyOf(policy) : Map.of();
    }
}
